### PR TITLE
WIP: VariableFunctions outside of SessionState Capability

### PR DESCRIPTION
**Pull request by romero126**
## PR Summary
The Proposed change Allows for the capability to specify a specific SessionState to Variable Cmdlet's

Example:
```
    $Module = New-Module -Name "TestVariables" {
        $script:TestGetVariable = "True"
        $script:TestSetVariable = "Set"
        $script:TestClearVariable = "ShouldBeBlank"
        $Script:TestRemoveVariable = "Should Not Exist"

        write-host "Finished Loading Module"
    }
    $Module = Import-Module $Module -PassThru
    New-Variable -SessionStateEntry $Module.SessionState -Name "TestNewVariable" -Value "True"
    Set-Variable -SessionStateEntry $Module.SessionState -Name "TestSetVariable" -Value "True"
    Clear-Variable -SessionStateEntry $Module.SessionState -Name "TestClearVariable"
    Remove-Variable -SessionStateEntry $Module.SessionState -Name "TestRemoveVariable"
    Get-Variable -SessionStateEntry $Module.SessionState -Name "Test*"
```

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Change Implements a new Parameter to Get/Set/Remove/Clear/New Variable to allow for retrieval of variables outside of current SessionState. 
    -

- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
   - Change is specific to Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs
   - Updates Helper Functions to allow for specific SessionState Functions.
   - 

- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
   - Change passes all of the MS Included Pester Tests
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - This PR is complete and should not be a breaking change. 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed]
(https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 1]
```diff
@@ -332,6 +369,23 @@ public string[] Exclude
...
+            {
+                return _sessionState;
+            }
+
+            set
+            {
+                _sessionState = value;
+            }
+        }
+        private SessionState _sessionState;
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: Would it make more sense to declare this as an *inherited param* from VariableCommandBase? (In which case you'd stick the definition up around lines 24-30.)

Might simplify the definitions a bit. ^^

**Review by vexx32 — COMMENTED** [review]
Looks pretty good, but might be an opportunity to simplify the implementation without losing anything? 😄

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs (hunk 1 above)
Comment (romero126) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: Placing repetitious code into VariableCommandBase works well however the SessionStateEntry is a Parameter. Parameters should stay with its relative function so we don't confuse/obfuscate the code base. Also isn't documentation generated from its <Summary> Tags? Then the parameter would become UnDocumented in get-help which is not pretty.

**Comment by TravisEz13** [comment]
There were some significant problems with your branch.  I manually rebased your branch and removed some duplicate commits which made the PR hard to review and you had added some files which were not part of the repo any more back, I removed those from your commits.

**Comment by TravisEz13** [comment]
Well... I thought I had...

**Comment by TravisEz13** [comment]
Okay, the rebase is done now.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 2]
```diff
@@ -1335,10 +1533,24 @@ private PSVariable ClearValue(PSVariable matchingVariable)
...
+        /// <param name="matchingVariable">
+        /// The variable that matched the name parameter(s).
+        /// </param>
+        ///
+        private PSVariable ClearValue(PSVariable matchingVariable)
+        {
+            return ClearValue(matchingVariable, SessionState);
+        }
+    } // ClearVariableCommand
+}
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: Missing a new line at the end of the file.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 3]
```diff
@@ -29,8 +32,9 @@ public abstract class VariableCommandBase : PSCmdlet
         #endregion parameters
 
         /// <summary>
-        /// The Include parameter for all the variable commands.
+        /// The Include parameter for all the variable commands
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: missing the period at the end of the comment

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 4]
```diff
@@ -50,15 +54,15 @@ protected string[] IncludeFilters
         private string[] _include = new string[0];
 
         /// <summary>
-        /// The Exclude parameter for all the variable commands.
+        /// The Exclude parameter for all the variable commands
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: missing period at the end of the comment (this is used for documentation.)

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 5]
```diff
@@ -70,30 +74,43 @@ protected string[] ExcludeFilters
         }
         private string[] _exclude = new string[0];
 
-        #region helpers
 
+        #region helpers
+
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: multiple blank lines in a row.  Please only one.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs [hunk 6]
```diff
@@ -450,31 +530,53 @@ public SwitchParameter PassThru
         }
         private bool _passThru;
 
+        /// <summary>
+        /// Specify the SessionState to get variable from.
+        /// </summary>
+        [Parameter]
+        public SessionState SessionStateEntry
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: Why cant this just be
```C#
        public SessionState SessionStateEntry
        {
            get;
             set;
        }
```

Comment (TravisEz13) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs]: In all places this parameter is defined...

**Comment by TravisEz13** [comment]
cc @PaulHigin

**Comment by lzybkr** [comment]
This looks like an alternative to the obscure syntax of specifying a module:

```powershell
& (Get-Module $myModule) Set-Variable -Name Something -Value $True
```

This current syntax should be disallowed in restricted language mode, and hence I believe this new parameter should also not work in restricted language mode.

**Comment by vexx32** [comment]
@lzybkr are you referring to `StrictMode` or `Constrained Language Mode`?

**Comment by romero126** [comment]
So my question is wouldn't we only want it to run correctly in FullLanguage mode?  ConstrainedLanguage, NoLanguage,RestrictedLanguage based on the descriptions all of them make restrictions on the Language in some way shape or form.

**Comment by PaulHigin** [comment]
I agree with @lzybkr that this needs to work correctly with language modes.  Unfortunately I haven't had a chance to look at this closely yet, and probably won't until next week.  At the very least we should postpone this change until @daxian-dbw untrusted input tracking work is merged.  Then this change can follow that pattern to conform to language modes.

**Comment by romero126** [comment]
I Agree that we need to make certain LanguageMode is considered and handled carefully Without introducing redundancy in code.

If there is a way to pull the most current sessionstate. I believe the best solution would be to create a SetLanguageModeAttribute() and just attach it to the Parameter it would be the best and cleanest implementation.

**Comment by JamesWTruher** [comment]
@romero126 I'm wondering if this isn't just another "scope", and whether the scope parameter should be made smarter to accept the module state in addition to what it does today?
Even if we take the behavior as submitted, with a change of such impact, we really need to see a set of tests which validate the new behavior.

**Comment by romero126** [comment]
@JamesWTruher This is a great idea. I really do like the idea putting this into the scope argument. Let me see what I can come up with, to implement this idea.

As well as working on getting some good pester tests in place.

**Pull request closed without merging**